File: backend/app/services/analytics_service.py
from sqlalchemy.orm import Session
from sqlalchemy import func, desc, select
from app.models.product import Product
from app.models.sale_item import SaleItem

def get_top_products(db: Session, limit: int = 5):
    """
    Returns top N best-selling products using:
    JOIN, GROUP BY, SUM, ORDER BY
    """
    stmt = (
        select(
            Product.name.label("product_name"),
            func.sum(SaleItem.quantity).label("total_sold")
        )
        .join(SaleItem, Product.id == SaleItem.product_id)
        .group_by(Product.name)
        .order_by(desc("total_sold"))
        .limit(limit)
    )
    
    results = db.execute(stmt).all()
    
    # Convert to list of dicts for JSON response
    return [{"product_name": r.product_name, "total_sold": int(r.total_sold)} for r in results]

File: backend/node_server/src/services/analyticsService.js
const { Sale } = require('../models');

/**
 * Returns top N best-selling products by total quantity sold.
 * Uses MongoDB aggregation pipeline.
 */
async function getTopProducts(limit = 5) {
  const results = await Sale.aggregate([
    { $unwind: '$items' },
    {
      $group: {
        _id: '$items.product_name',
        total_sold: { $sum: '$items.quantity' },
      },
    },
    { $sort: { total_sold: -1 } },
    { $limit: limit },
    {
      $project: {
        _id: 0,
        product_name: '$_id',
        total_sold: 1,
      },
    },
  ]);

  return results;
}

/**
 * Returns products with stock <= 5 (low-stock alerts)
 */
async function getLowStockProducts() {
  const { Product } = require('../models');
  return Product.find({ stock: { $lte: 5 } }).sort({ stock: 1 });
}

module.exports = { getTopProducts, getLowStockProducts };
